src/test/scala/com/microsoft/hyperspace/index/BucketUnionTest.scala

```diff
@@ -0,0 +1,134 @@
+
+    intercept[IllegalArgumentException] {
+      BucketUnion(
+        Seq(df1.queryExecution.optimizedPlan, df2.queryExecution.optimizedPlan),
+        BucketSpec(1, Seq(), Seq()))
+      fail("shouldn't be reached")
+    }
+    intercept[IllegalArgumentException] {
+      BucketUnion(
+        Seq(df1.queryExecution.optimizedPlan, df1_2.queryExecution.optimizedPlan),
+        BucketSpec(1, Seq(), Seq()))
+      fail("shouldn't be reached")
+    }
+    BucketUnion(
+      Seq(df1.queryExecution.optimizedPlan, df1_1.queryExecution.optimizedPlan),
+      BucketSpec(1, Seq(), Seq()))
+  }
+
+  test("BucketUnionStrategy test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val bucket = BucketUnion(
+      Seq(df1.queryExecution.optimizedPlan, df1_1.queryExecution.optimizedPlan),
+      BucketSpec(1, Seq(), Seq()))
+
+    var bucketUnionExecSeen = false
+    BucketUnionStrategy(bucket).collect {
+      case BucketUnionExec(_, _) => bucketUnionExecSeen = true
+    }
+    assert(bucketUnionExecSeen)
+
+    bucketUnionExecSeen = false
+    BucketUnionStrategy(df1.queryExecution.optimizedPlan).collect {
+      case BucketUnionExec(_, _) => bucketUnionExecSeen = true
+    }
+    assert(!bucketUnionExecSeen)
+  }
+
+  test("BucketUnionExec test") {
+    import spark.implicits._
+    val df1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val p1 = df1.repartition(10)
+    val df1_1 = Seq((1, "name1"), (2, "name2")).toDF("id", "name")
+    val p1_1 = df1_1.repartition(9)
+    val p1_2 = df1_1.repartition(10)
+
+    intercept[AssertionError] {
+      val bucket = BucketUnion(
+        Seq(p1.queryExecution.optimizedPlan, p1_1.queryExecution.optimizedPlan),
+        BucketSpec(10, Seq(), Seq()))
+      spark.sessionState.executePlan(bucket).sparkPlan
+      fail("shouldn't be reached")
+    }
+
+    val bucket = BucketUnion(
+      Seq(p1.queryExecution.optimizedPlan, p1_2.queryExecution.optimizedPlan),
+      BucketSpec(10, Seq(), Seq()))
+
+    val exec = BucketUnionStrategy(bucket)
+
+    val plan = exec.collect {
+      case p: BucketUnionExec =>
+        assert(p.bucketSpec.numBuckets == 10)
+        assert(p.children.length == 2)
+        assert(p.output.length == p1.schema.fields.length)
+    }
+  }
+
+  test("BucketUnionRDD test") {
+    import spark.implicits._
+    val df1 = Seq((2, "name1"), (3, "name2")).toDF("id", "name")
+    val p1 = df1.repartition(10, $"id")
+    val df1_1 = Seq((2, "name1"), (3, "name2")).toDF("id", "name")
+    val p1_1 = df1_1.repartition(10, $"id")
+    val bucketSpec = BucketSpec(10, Seq("id"), Seq())
+
+    val rdd = new BucketUnionRDD[Row](spark.sparkContext, Seq(p1.rdd, p1_1.rdd), bucketSpec)
+    assert(rdd.getPartitions.length == 10)
+    assert(rdd.collect.length == 4)
+    rdd.partitions.head.asInstanceOf[BucketUnionRDDPartition]
```

Comment:
what is this line?

Reply:
did you mean `assert(rdd.partitions.head.isInstanceOf[BucketUnionRDDPartition])`?

Reply:
Yes asInstanceOf throws an exception for a failure but I'll change to isInstanceOf.
Thanks!